src/smartptrs.h

```diff
@@ -7,80 +7,85 @@
-  virtual bool IsCpuAccessible() const = 0;
-  virtual void GetOnCpu() = 0;  // Allocates p_cpu_ if necessary and copies p_device_ memory into it
-  virtual void CopyFromDevice(size_t begin_dest, DeviceMemoryBase& source, size_t begin_source, size_t size_in_bytes) = 0;
+
+  virtual void AllocateCpu() = 0;      // Allocates p_cpu_ if necessary (using appropriate memory type for interop)
+  virtual void CopyDeviceToCpu() = 0;  // Allocates p_cpu_ if necessary and copies p_device_ memory into it
+  virtual void CopyCpuToDevice() = 0;
+  virtual void CopyFrom(size_t begin_dest, DeviceBuffer& source, size_t begin_source, size_t size_in_bytes) = 0;
 
   uint8_t* p_device_{};
   uint8_t* p_cpu_{};
   size_t size_in_bytes_{};
 };
 
+// A DeviceSpan is how a DeviceBuffer is used. It can be thought of as a std::span for device memory with
+// utilities to interop with CPU memory. It is what Allocate<T> returns and what should be passed around by value.
 template <typename T>
-struct DeviceMemorySpan;
+struct DeviceSpan {
+  DeviceSpan() = default;
+  DeviceSpan(std::shared_ptr<DeviceBuffer>&& memory)
+      : p_device_memory_{std::move(memory)}, begin_{}, length_{p_device_memory_->size_in_bytes_ / sizeof(T)} {}
 
-template <typename T>
-struct DeviceMemory : DeviceMemoryBase {
-  size_t Size() const { return size_in_bytes_ / sizeof(T); }
-  std::span<T> DeviceSpan() { return std::span<T>(reinterpret_cast<T*>(p_device_), Size()); }
-  std::span<T> CpuSpan() {
-    if (!p_cpu_) GetOnCpu();
-    return std::span<T>(reinterpret_cast<T*>(p_cpu_), Size());
-  }
+  bool empty() const { return length_ == 0; }
+  size_t size() const { return length_; }
 
-  DeviceMemorySpan<T> subspan() { return DeviceMemorySpan<T>(*this, 0, Size()); }
-  DeviceMemorySpan<T> subspan(size_t begin, size_t length) { return DeviceMemorySpan<T>(*this, begin, length); }
-  DeviceMemorySpan<T> subspan_cpu(std::span<T> span) { return DeviceMemorySpan<T>(*this, span.data() - CpuSpan().data(), span.size()); }
-  DeviceMemorySpan<T> subspan_device(std::span<T> span) { return DeviceMemorySpan<T>(*this, span.data() - DeviceSpan().data(), span.size()); }
+  DeviceSpan<T> subspan(size_t begin, size_t length) { return DeviceSpan<T>(*p_device_memory_, begin_ + begin, length); }
 
- private:
-  static void CheckSize() {
-    // Ensure we're the same size as DeviceMemory so we can be cast to it
-    // Has to be in a method because sizeof(TDeviceMemory) is not known outside of methods
-    static_assert(sizeof(DeviceMemory) == sizeof(DeviceMemoryBase));
+  // Return the device accessible memory. Should only be done in device specific code, as it's not CPU accessible
+  std::span<T> Span() { return std::span<T>{reinterpret_cast<T*>(p_device_memory_->p_device_) + begin_, length_}; }
+
+  // Return the CPU accessible memory, allocating if necessary (note, to get the current device memory on CPU, use 'CopyDeviceToCpu' instead)
+  std::span<T> CpuSpan() {
+    p_device_memory_->AllocateCpu();
```

Comment:
I tried doing this to match RoamingArray but it winds up not having enough information to know what memory is valid or what the intention is, and wound up having more functions for the opposite reason.

If you're going to totally overwrite the memory from the CPU side, you don't want a device->CPU copy, but you do want a CPU->device copy after you're done.
If you only want to read the device contents, you want a device->CPU copy, but no CPU->device copy afterwards.
If the memory is aliased somewhere else (like in an OrtValue), the DeviceSpan won't ever know when you try to access it next, so it can't auto copy it back.
I started trying to do this, but it needed 'CpuSpan()' for read only cases, 'MutableCpuSpan()' for needing to write back, and flags to say what memory is valid, and it just got messy and still wasn't perfect for the uninitialized memory case. This way felt simpler and explicit. We know exactly when we're doing a copy.